helix-term/src/commands.rs

```diff
@@ -5462,63 +5475,73 @@ fn select_textobject(cx: &mut Context, objtype: textobject::TextObject) {
+                    if ch.is_ascii_alphanumeric() {
+                        return range;
                     }
-                });
-                doc.set_selection(view.id, selection);
-            };
-            textobject(cx.editor);
-            cx.editor.last_motion = Some(Motion(Box::new(textobject)));
-        }
-    });
+                    // TODO: cancel new ranges if inconsistent surround matches across lines
+                    textobject::textobject_surround(text, range, objtype, ch, count)
+                }
+            }
+        });
 
-    if let Some((title, abbrev)) = match objtype {
-        textobject::TextObject::Inside => Some(("Match inside", "mi")),
-        textobject::TextObject::Around => Some(("Match around", "ma")),
-        _ => return,
-    } {
-        let help_text = [
-            ("w", "Word"),
-            ("W", "WORD"),
-            ("c", "Class (tree-sitter)"),
-            ("f", "Function (tree-sitter)"),
-            ("p", "Parameter (tree-sitter)"),
-            ("m", "Matching delimiter under cursor"),
-            (" ", "... or any character acting as a pair"),
-        ];
+        doc.set_selection(view.id, selection);
+    };
+    textobject(cx.editor);
+    cx.editor.last_motion = Some(Motion(Box::new(textobject)));
+}
 
-        cx.editor.autoinfo = Some(Info::new(
-            title,
-            help_text
-                .into_iter()
-                .map(|(col1, col2)| (col1.to_string(), col2.to_string()))
-                .collect(),
-        ));
-        cx.editor.pseudo_pending = Some(abbrev.to_string());
+macro_rules! select_textobject_commands {
+    ( $($name:ident($objtype:ident, $objsel:ident($val:expr));)* ) => {
+        $(
+            fn $name(cx: &mut Context) {
+                select_textobject(
+                    cx,
+                    textobject::TextObject::$objtype,
+                    TextObjectSel::$objsel($val),
+                )
+            }
+        )*
     };
 }
 
+select_textobject_commands! {
+    select_around_word(Around, Word(false));
+    select_inside_word(Inside, Word(false));
+    select_around_long_word(Around, Word(true));
+    select_inside_long_word(Inside, Word(true));
+    select_around_class(Around, Treesitter("class"));
+    select_inside_class(Inside, Treesitter("class"));
+    select_around_function(Around, Treesitter("function"));
+    select_inside_function(Inside, Treesitter("function"));
+    select_around_parameter(Around, Treesitter("parameter"));
+    select_inside_parameter(Inside, Treesitter("parameter"));
+    select_around_cursor_pair(Around, Matching(None));
+    select_inside_cursor_pair(Inside, Matching(None));
+}
+
+fn select_inside_pair(cx: &mut Context) {
+    if let Some(ch) = cx.last_key.and_then(|ev| ev.char()) {
+        select_textobject(
+            cx,
+            textobject::TextObject::Inside,
+            TextObjectSel::Matching(Some(ch)),
+        )
+    }
+}
```

Comment:
How about this? When the command is run via command palette (or presumably non-fallback binding), it prompts for a key, which I believe is the desired behavior.

Reply:
That is what we want, but putting a `last_key` field on the Context feels hacky and unnecessary. Right now it's required because the fallback command is also defined to be a `MappableCommand`, and therefore it has no idea which key to use when it's invoked as a fallback. This can be easily solved by adding a new `FallbackKeyCommand` type as I mentioned above, which also takes a KeyEvent as a parameter.

Reply:
I think I get it now, but I also just realized that when run from the command palette, the `on_next_key` callback never actually gets called; I'm guessing it gets discarded as a result of how the picker works. I expect to be able to figure this out with a little more research.

Reply:
Does this work?